### PR TITLE
[usage] Add a placeholder Stripe webhook

### PR DESCRIPTION
**Pull request by andrew-farries**
## Description

Add a placeholder [Stripe webhook](https://stripe.com/docs/webhooks) to the usage component. 

In later PRs, the webhook will receive `invoice.finalized` events from Stripe so that we can update instance usage records to show that they have been included in a particular invoice.

## Related Issue(s)

Part of #9036 and https://github.com/gitpod-io/gitpod/issues/10937

## How to test

* Install the[ `stripe` CLI](https://stripe.com/docs/stripe-cli) into the workspace and run `stripe login`.
* Port forward to the database in the preview environment:
```
kubectl port-forward svc/mysql 3306:3306
```
* Run the `usage` component locally:
```
DB_USERNAME=gitpod DB_HOST=localhost DB_PORT=3306 DB_PASSWORD=<>  go run . run  
```
* Forward `stripe` events to the local webhook endpoint:
```
stripe listen --skip-verify --forward-to localhost:9002/webhook
```
* Trigger a series of Stripe events:
```
stripe trigger invoice.finalized
```
This should start a series of Stripe events (setting up a customer and payment methods), each of which should be handled with a `200 OK`  by the webhook:

image

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by andrew-farries** [comment]
I've put the webhook on the `usage` component which means we will need to set up ingress (via `proxy`) to allow the `usage` component to be reachable by Stripe. 

An alternative, would be to put the hook on `server` as we already have other webhook handling there. `usage` feels like the right place for this though.

File: components/usage/pkg/server/server.go
Comment (easyCZ) [inline components/usage/pkg/server/server.go:141]: I'd recommend prefixing this with something like `/stripe/invoices/webhook`. This makes it easier to debug when you see just a request HTTP path without knowing what it does.

File: components/usage/pkg/stripe/webhook.go
Comment (easyCZ) [inline components/usage/pkg/stripe/webhook.go:24]: I believe there' a middleware for this

File: components/usage/pkg/stripe/webhook.go
Comment (easyCZ) [inline components/usage/pkg/stripe/webhook.go:39]:
```
dec := json.Unmarshaller(http.Body)
if err := dec.Unmarshal(&event); err != nil { ... }
```

This should work. You shouldn't need to read all of the body into memory first and only then start unmarshalling. By using the above, you can stream unmarshal and avoid some buffering.

File: components/usage/pkg/stripe/webhook.go
Comment (easyCZ) [inline components/usage/pkg/stripe/webhook.go:23]: Also worth wrapping in middleware which rejects non `application/json` content type.

**Comment by easyCZ** [comment]
> I've put the webhook on the usage component which means we will need to set up ingress (via proxy) to allow the usage component to be reachable by Stripe.

I'd avoid this. We don't have AuthN on the usage component so it shouldn't be directly exposed to the internet.

> An alternative, would be to put the hook on server as we already have other webhook handling there. usage feels like the right place for this though.

Do this, but proxy from Server to usage - server becomes the authN ingress, but you can keep logic in usage.

**Comment by easyCZ** [comment]
Thinking about this, I'd actually implement the Usage part as a gRPC method which the `server` would delegate. I think in practice all we would need is something like `BillingService.FinalizeInvoice(..)`

**Comment by andrew-farries** [comment]
> > I've put the webhook on the usage component which means we will need to set up ingress (via proxy) to allow the usage component to be reachable by Stripe.
> 
> I'd avoid this. We don't have AuthN on the usage component so it shouldn't be directly exposed to the internet.

Without AuthN on the `usage` gRPC methods, we could still expose just the webhook though right?

My preference for putting the webhook on the `usage` component comes simply from not wanting to load yet more functionality onto `server`.

File: components/usage/pkg/server/server.go
Comment (andrew-farries) [reply components/usage/pkg/server/server.go:141]: b0d3d891b

File: components/usage/pkg/stripe/webhook.go
Comment (andrew-farries) [reply components/usage/pkg/stripe/webhook.go:23]: I couldn't find any middleware that does exactly this. Did you have something in mind?

File: components/usage/pkg/stripe/webhook.go
Comment (andrew-farries) [reply components/usage/pkg/stripe/webhook.go]: 766430278

File: components/usage/pkg/stripe/webhook.go
Comment (andrew-farries) [reply components/usage/pkg/stripe/webhook.go]: 354cb8351

**Comment by easyCZ** [comment]
> Without AuthN on the usage gRPC methods, we could still expose just the webhook though right?

That's fair, my worry is the added complexity where you now have to reason through some parts of it being public, while others not. It increases the threat surface for the system but worse it causes extra cognitive load on the engineers. The current (unfavourable) case where only `server` is exposed at least keeps the assumptions constant

**Review by easyCZ — APPROVED** [review]
I'm okay to go with this for now, to unblock you on other PRs but I'd like to understand how others feel about this congitive overhead as they will be the ones on-call for this.

/hold

**Comment by andrew-farries** [comment]
Closing, in favour of https://github.com/gitpod-io/gitpod/pull/11806 that we've agreed that the public-api is a better home for the webhook

**Pull request closed without merging**
